src/intellisense/texDocSymbolProvider.ts

```diff
@@ -0,0 +1,366 @@
+                            ...attributes
+                        };
+                    }
+                    break;
+            }
+            break;
+
+        case 'mathenv':
+            switch (node.env) {
+                case 'string':
+                    element = {
+                        type: TeXElementType.Environment,
+                        name: node.env,
+                        label: `${node.env.charAt(0).toUpperCase()}${node.env.slice(1)}`,
+                        ...attributes
+                    };
+                    break;
+            }
+            break;
+    }
+
+    if (element !== undefined) {
+        root.children.push(element);
+        root = element;
+    }
+
+    if ('content' in node && typeof node.content !== 'string') {
+        for (const sub of node.content) {
+            if (['string', 'parbreak', 'whitespace'].includes(sub.type)) {
+                continue;
+            }
+            await parseNode(sub, root);
+        }
+    }
+}
+
+export class DocSymbolProvider implements vscode.DocumentSymbolProvider {
+    async provideDocumentSymbols(document: vscode.TextDocument): Promise<vscode.DocumentSymbol[]> {
+        const sections = await genTexElements(document);
+        return this.elementsToSymbols(sections);
+    }
+
+    private elementsTypeCast(section: TeXElement): vscode.SymbolKind {
+        switch (section.type){
+            case TeXElementType.Section:
+            case TeXElementType.SectionAst:
+                return vscode.SymbolKind.Struct;
+            case TeXElementType.Environment:
+                return vscode.SymbolKind.Package;
+            case TeXElementType.Command:
+                return vscode.SymbolKind.Number;
+            case TeXElementType.SubFile:
+                return vscode.SymbolKind.File;
+            case TeXElementType.BibItem:
+                return vscode.SymbolKind.Class;
+            case TeXElementType.BibField:
+                return vscode.SymbolKind.Constant;
+            default:
+                return vscode.SymbolKind.String;
+        }
+    }
+
+    private elementsToSymbols(sections: TeXElement[]): vscode.DocumentSymbol[] {
+        const symbols: vscode.DocumentSymbol[] = [];
+        sections.forEach(section => {
+            const range = new vscode.Range(section.lineFr, 0, section.lineTo, 65535);
+            const symbol = new vscode.DocumentSymbol(
+                section.label || 'empty', '',
+                this.elementsTypeCast(section),
+                range, range);
+            symbols.push(symbol);
+            if (section.children.length > 0) {
+                symbol.children = this.elementsToSymbols(section.children);
+            }
+        });
+        return symbols;
+    }
+
+    get triggers(){
+        const latexSelector = ['latex', 'latex-expl3', 'pweave', 'jlweave', 'rsweave'].map((id) => {
+            return { language: id };
```

Comment:
also restrict the scheme to ROOT_NAME

see reference above

Reply:
OK